conjureup/controllers/clouds/common.py

```diff
@@ -0,0 +1,38 @@
+import asyncio
+
+from conjureup import events, utils
+from conjureup.app_config import app
+from conjureup.models.provider import LocalhostError, LocalhostJSONError
+
+
+class BaseCloudController:
+    retry_count = 0
+    max_retry = 20
+
+    async def _monitor_localhost(self, cb):
+        """ Checks that localhost/lxd is available and listening,
+        updates widget accordingly
+        """
+        if events.LXDAvailable.is_set():
+            return
+
+        try:
+            await app.provider.is_server_available()
+            events.LXDAvailable.set()
+            self.retry_count = 0
+            cb()
+        except (LocalhostError,
+                LocalhostJSONError):
+            if self.retry_count == self.max_retry:
+                raise
+            app.provider._set_lxd_dir_env()
+            self.retry_count += 1
+            await asyncio.sleep(2)
+            await self._monitor_localhost(cb)
```

Comment:
Recursion seems like an odd way to do a retry loop.  I think a `for` loop would be easier to follow (it took me a second to figure out that this was a retry loop).  Actually, why even have a max retry?  If the user sits on the cloud selection screen for too long it would just stop working.  I think a cancel event (using `run_with_interrupt`) would be better, so that the task can just keep trying for as long as needed and then be cancelled cleanly when a cloud is selected.  Something like:

```python
while not self.cancel_monitor.is_set():
    try:
        await app.provider.is_server_available()
        events.LXDAvailable.set()
        cb()
        return
    except (LocalhostError, LocalhostJSONError):
        await utils.run_with_interrupt(asyncio.sleep(2), self.cancel_monitor)
    except FileNotFoundError:
        if app.headless:
            utils.error("Unable to find lxc executable, please make "
                        "sure LXD is installed: `sudo snap install lxd`")
            events.Shutdown.set(1)
        await utils.run_with_interrupt(asyncio.sleep(5), self.cancel_monitor)
```